storage/src/backend/http_proxy.rs

```diff
@@ -0,0 +1,541 @@
+/// - `HEAD /path/to/blob` to get the blob size
+/// - `GET /path/to/blob` to read the blob
+///
+/// The http proxy server should respect [the `Range` header](https://www.rfc-editor.org/rfc/rfc9110.html#name-range) to support range reading.
+pub struct HttpProxy {
+    addr: String,
+    path: String,
+    client: Client,
+    metrics: Option<Arc<BackendMetrics>>,
+}
+
+/// HttpProxyReader is a BlobReader to implement the HttpProxy backend driver.
+pub struct HttpProxyReader {
+    client: Client,
+    uri: Uri,
+    metrics: Arc<BackendMetrics>,
+}
+
+#[derive(Clone)]
+struct LocalClient {
+    client: Arc<HyperClient<UnixConnector>>,
+    runtime: Arc<Runtime>,
+}
+
+#[derive(Clone)]
+enum Client {
+    Local(LocalClient),
+    Remote(Arc<Connection>),
+}
+
+enum Uri {
+    Local(Arc<hyper::Uri>),
+    Remote(String),
+}
+
+fn range_str_for_header(offset: u64, len: Option<usize>) -> String {
+    match len {
+        Some(len) => format!("bytes={}-{}", offset, offset + len as u64 - 1),
+        None => format!("bytes={}-", offset),
+    }
+}
+
+fn build_tokio_runtime(name: &str, thread_num: usize) -> Result<Runtime> {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name(name)
+        .worker_threads(thread_num)
+        .enable_all()
+        .build()?;
+    Ok(runtime)
+}
+
+impl LocalClient {
+    async fn do_req(
+        &self,
+        uri: Arc<hyper::Uri>,
+        only_head: bool,
+        offset: u64,
+        len: Option<usize>,
+    ) -> BackendResult<Response<Body>> {
+        let method = if only_head { Method::HEAD } else { Method::GET };
+        let req = Request::builder()
+            .method(method)
+            .uri(uri.as_ref())
+            .header(http::header::RANGE, range_str_for_header(offset, len))
+            .body(Body::default())
+            .map_err(HttpProxyError::BuildHttpRequest)?;
+        let resp = self
+            .client
+            .request(req)
+            .await
+            .map_err(HttpProxyError::LocalRequest)?;
+        Ok(resp)
+    }
+
+    fn get_headers(&self, uri: Arc<hyper::Uri>) -> BackendResult<HeaderMap<HeaderValue>> {
+        let headers = self
+            .runtime
+            .block_on(self.do_req(uri, true, 0, None))?
+            .headers()
+            .to_owned();
+        Ok(headers)
+    }
+
+    fn try_read(&self, uri: Arc<hyper::Uri>, offset: u64, len: usize) -> BackendResult<Vec<u8>> {
+        self.runtime.block_on(async {
+            let resp = self.do_req(uri, false, offset, Some(len)).await;
+            match resp {
+                Ok(resp) => body::to_bytes(resp)
+                    .await
+                    .map_err(|e| HttpProxyError::ReadResponseBody(e).into())
+                    .map(|bytes| bytes.to_vec()),
+                Err(e) => Err(e),
+            }
+        })
+    }
+}
+
+impl BlobReader for HttpProxyReader {
+    // Note! If the blob is compressed, the function will always return 0
```

Comment:
Is this comment useful?

Reply:
Yes, it's useful. If the blobs are comporessed, we cannnot get the real blob size before we complete the stream reading.

Here is an example in nydus-snapshotter https://github.com/containerd/nydus-snapshotter/blob/3d37ee2f6834ca1626abcd69063f96e1d69d397f/pkg/converter/convert_unix.go#L233.